File: src/main/java/pl/patrykdolata/lifemanager/mapper/UserMapper.java
package pl.patrykdolata.lifemanager.mapper;

import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.StringUtils;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Named;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import pl.patrykdolata.lifemanager.domain.UserEntity;
import pl.patrykdolata.lifemanager.model.NewUser;

import java.util.Date;

@Mapper(componentModel = "spring", imports = {RandomStringUtils.class, Date.class})
public abstract class UserMapper {

    public static final int KEY_LENGTH = 20;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Mapping(target = "password", source = "password", qualifiedByName = {"mapPassword"})
    @Mapping(target = "activated", constant = "false")
    @Mapping(target = "activationKey", expression = "java(RandomStringUtils.randomNumeric(KEY_LENGTH))")
    @Mapping(target = "createdAt", expression = "java(new Date().getTime())")
    public abstract UserEntity toEntity(NewUser newUser);

    @Named("mapPassword")
    protected String mapPassword(String password) {
        if (StringUtils.isEmpty(password)) {
            return passwordEncoder.encode(RandomStringUtils.randomAlphanumeric(KEY_LENGTH));
        }
        return passwordEncoder.encode(password);
    }
}

File: src/main/kotlin/pl/patrykdolata/lifemanager/mapper/UserMapper.kt
package pl.patrykdolata.lifemanager.mapper

import org.apache.commons.lang3.RandomStringUtils
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.Mappings
import org.mapstruct.Named
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.crypto.password.PasswordEncoder
import pl.patrykdolata.lifemanager.domain.AuthorityEntity
import pl.patrykdolata.lifemanager.domain.UserEntity
import pl.patrykdolata.lifemanager.model.NewUser
import java.util.*

@Mapper(componentModel = "spring", imports = [Date::class, RandomStringUtils::class])
abstract class UserMapper {

    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    companion object {
        const val KEY_LENGTH = 20
    }

    @Mappings(
            Mapping(target = "id", ignore = true),
            Mapping(target = "password", source = "newUser.password", qualifiedByName = ["mapPassword"]),
            Mapping(target = "activated", constant = "true"),
            Mapping(target = "activationKey", expression = "java(RandomStringUtils.randomNumeric(KEY_LENGTH))"),
            Mapping(target = "createdAt", expression = "java(new Date().getTime())"),
            Mapping(target = "resetKey", ignore = true),
            Mapping(target = "authorities", source = "isAdmin", qualifiedByName = ["getAuthorities"])
    )
    abstract fun toEntity(newUser: NewUser, isAdmin: Boolean = false): UserEntity

    @Named("mapPassword")
    fun mapPassword(password: String): String {
        return when (password.isEmpty()) {
            true -> passwordEncoder.encode(RandomStringUtils.randomAlphanumeric(KEY_LENGTH))
            false -> passwordEncoder.encode(password)
        }
    }

    @Named("getAuthorities")
    fun getAuthorities(isAdmin: Boolean): Set<AuthorityEntity> {
        val userAuthority = AuthorityEntity("USER")
        return if (isAdmin) return setOf(userAuthority, AuthorityEntity("ADMIN")) else setOf(userAuthority)
    }
}
